### ISSUE TITLE
event not reported in epoll edge-trigger mode after mod operation

### ISSUE THREAD

**Issue by bobbyvinon** [shadow/shadow#343]
i have some server code, running into shadow plugin, using epoll to work with sockets.
server receive request and answers to it:

```
  // sending answer
  int s = write(session->fd, answer, strlen(answer));

  // setting EPOLLOUT
  struct epoll_event event;

  event.events = EPOLLOUT | EPOLLET;
  event.data.fd = session->fd;

  epoll_ctl(epollfd, EPOLL_CTL_MOD, fd, &event);
```

after answer was sent, epoll signals session->fd EPOLLOUT flag and i close connection.
this code works fine on baremetal, but on shadow, i didn't see this EPOLLOUT after data was sent.
(in log i see "status changed in epoll 13 for descriptor 16", but epoll_wait return nothing).

if i replace:
```
  epoll_ctl(epollfd, EPOLL_CTL_MOD, fd, &event);
```

with:
```
  epoll_ctl(epollfd, EPOLL_CTL_DEL, fd, NULL);
  epoll_ctl(epollfd, EPOLL_CTL_ADD, fd, &event);
```

code works fine in shadow.

**Comment by robgjansen** [comment]
In Shadow, an event for a file descriptor in edge-triggered mode is only triggered if the read/write status of that descriptor has changed. It does not change when packets are sent/received, unless that changes the read/write status of the descriptor. Are you suggesting that it should work differently?

The example above is hard to follow because it is missing in which order the EPOLL_CTL_ADD was initially called (was it called before the write() or after?).

This would be much easier to get this correct in Shadow if we had a test case for it. Would you be willing to add a new function test case:
https://github.com/shadow/shadow/blob/master/src/test/epoll/shd-test-epoll.c
Running `./setup test` would then indicate that the test fails in Shadow but passes outside of Shadow.

Thanks!

**Comment by bobbyvinon** [comment]
after incoming connection received, this code was called:

int connfd = accept(listen_sd, (struct sockaddr *)&clientaddr, &clilen);
make_socket_non_blocking(connfd);
add_epoll_event(epoll_fd, connfd);

after this i didn't call read() or write() - just waiting for EPOLLIN.
so EPOLL_CTL_ADD called before write(), and EPOLL_CTL_MOD called after write.
on linux this code works without a problems.

int make_socket_non_blocking(int sfd)
{
  int flags, s;

  flags = fcntl(sfd, F_GETFL, 0);
  if (flags == -1)
  {
    perror("fcntl");
    return -1;
  }

  flags |= O_NONBLOCK;
  s = fcntl(sfd, F_SETFL, flags);
  if (s == -1)
  {
    perror("fcntl");
    return -1;
  }

  return 0;
}

void add_epoll_event(int efd, int sfd)
{
  struct epoll_event event;
  memset(&event, 0, sizeof(event));
  event.data.fd = sfd;
  event.events = EPOLLIN | EPOLLET;
  int s = epoll_ctl(efd, EPOLL_CTL_ADD, sfd, &event);
  if (s == -1)
  {
    perror("epoll_ctl");
    abort();
  }
}
